src/linux.rs

```diff
@@ -0,0 +1,398 @@
+    }
+}
+
+pub fn get_num_physical_cpus() -> usize {
+    let file = match File::open("/proc/cpuinfo") {
+        Ok(val) => val,
+        Err(_) => return get_num_cpus(),
+    };
+    let reader = BufReader::new(file);
+    let mut map = HashMap::new();
+    let mut physid: u32 = 0;
+    let mut cores: usize = 0;
+    let mut chgcount = 0;
+    for line in reader.lines().filter_map(|result| result.ok()) {
+        let mut it = line.split(':');
+        let (key, value) = match (it.next(), it.next()) {
+            (Some(key), Some(value)) => (key.trim(), value.trim()),
+            _ => continue,
+        };
+        if key == "physical id" {
+            match value.parse() {
+                Ok(val) => physid = val,
+                Err(_) => break,
+            };
+            chgcount += 1;
+        }
+        if key == "cpu cores" {
+            match value.parse() {
+                Ok(val) => cores = val,
+                Err(_) => break,
+            };
+            chgcount += 1;
+        }
+        if chgcount == 2 {
+            map.insert(physid, cores);
+            chgcount = 0;
+        }
+    }
+    let count = map.into_iter().fold(0, |acc, (_, cores)| acc + cores);
+
+    if count == 0 {
+        get_num_cpus()
+    } else {
+        count
+    }
+}
+
+/// Cached CPUs calculated from cgroups.
+///
+/// If 0, check logical cpus.
+// Allow deprecation warnings, we want to work on older rustc
+#[allow(warnings)]
+static CGROUPS_CPUS: AtomicUsize = ::std::sync::atomic::ATOMIC_USIZE_INIT;
+
+fn cgroups_num_cpus() -> Option<usize> {
+    #[allow(warnings)]
+    static ONCE: Once = ::std::sync::ONCE_INIT;
+
+    ONCE.call_once(init_cgroups);
+
+    let cpus = CGROUPS_CPUS.load(Ordering::Acquire);
+
+    if cpus > 0 {
+        Some(cpus)
+    } else {
+        None
+    }
+}
+
+fn init_cgroups() {
+    // Should only be called once
+    debug_assert!(CGROUPS_CPUS.load(Ordering::SeqCst) == 0);
+
+    match load_cgroups("/proc/self/cgroup", "/proc/self/mountinfo") {
+        Some(quota) => {
+            if quota == 0 {
+                return;
+            }
+
+            let logical = logical_cpus();
+            let count = ::std::cmp::min(quota, logical);
+
+            CGROUPS_CPUS.store(count, Ordering::SeqCst);
+        }
+        None => return,
+    }
+}
+
+fn load_cgroups<P1, P2>(cgroup_proc: P1, mountinfo_proc: P2) -> Option<usize>
+where
+    P1: AsRef<Path>,
+    P2: AsRef<Path>,
+{
+    let subsys = some!(Subsys::load_cpu(cgroup_proc));
+    let mntinfo = some!(MountInfo::load_cpu(mountinfo_proc));
+    let cgroup = some!(Cgroup::mount(mntinfo, subsys));
+    cgroup.cpu_quota()
+}
+
+struct Cgroup {
+    base: PathBuf,
+}
+
+struct MountInfo {
+    root: String,
+    mount_point: String,
+}
+
+struct Subsys {
+    base: String,
+}
+
+impl Cgroup {
+    fn new(dir: PathBuf) -> Cgroup {
+        Cgroup {
+            base: dir,
+        }
+    }
+
+    fn mount(mntinfo: MountInfo, subsys: Subsys) -> Option<Cgroup> {
```

Comment:
There is no mount operation done here, `absolute_path` maybe?

Reply:
Oh, you mean the function name. I had originally called it `mounted`, is that better? Or perhaps `translate`?

Reply:
Oh yes, sorry for being a bit cryptic. `mounted` sounds fine.